Add joystick drone connector

sohampatkar10 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jhu-asco/aerial_autonomy/102)
<!-- Reviewable:end -->

garimellagowtham commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[include/aerial_autonomy/controller_hardware_connectors/joystick_velocity_controller_drone_connector.h, line 35 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7RPV15bdAjvIiaavQ:-Kt7RPV15bdAjvIiaavR:b-p429w9) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/controller_hardware_connectors/joystick_velocity_controller_drone_connector.h#L35)):*
> ```objc
>       Sensor<VelocityYaw> &velocity_sensor)
>       : ControllerHardwareConnector(controller, HardwareType::UAV),
>         drone_hardware_(drone_hardware), velocity_sensor_(velocity_sensor) {
> ```

Move implementations to cpp

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 43 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Ks8odr9oSo9I_S8y:-Kt7Ks8odr9oSo9I_S8z:b-qids8w) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/base_sensor.h#L43)):*
> ```objc
>   * @brief set sensor status
>   */
>   void setSensorStatus(SensorStatus sensor_status) {
> ```

why should external classes set sensor status? This function should not be public.

---

*[include/aerial_autonomy/sensors/guidance.h, line 18 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7LMaxFZgM7AQ-R93D:-Kt7LMaxFZgM7AQ-R93E:b8o5ria) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/guidance.h#L18)):*
> ```objc
>   *
>   */
>   Guidance(parsernode::Parser &drone_hardware)
> ```

set sensor status to valid always for guidance

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 23 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7QIGiUIPK_6ouwAqo:-Kt7QIGiUIPK_6ouwAqp:bg7pxp6) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L23)):*
> ```objc
>   * @param Config for velocity sensor
>   */
>   VelocitySensor(parsernode::Parser &drone_hardware, ros::NodeHandle nh,
> ```

Ideally, should be taking a GPSSensor as an option and check validity if the sensor is available. Add todo for now

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7MEZjb8EC94IwL9xp:-Kt7MEZjb8EC94IwL9xq:bfuyu1t) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L53)):*
> ```objc
> 
>     tf::Vector3 global_pos =
>         (sensor_tf * sensor_world_tf * sensor_tf.inverse()).getOrigin();
> ```

This is still wrong. The quadrotor takes the world frame aligned with NED (magnetic) of earth and with origin at where quadrotor started. So you need to premultiply with quads initial transformation when it started. In the constructor store the quads zero transformation(Assuming we start the node when quad is landed) and pre multiply your transformation

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 55 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7N8SvK6Zb02iCweT1:-Kt7N8SvK6Zb02iCweT2:b-a2yazf) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L55)):*
> ```objc
>         (sensor_tf * sensor_world_tf * sensor_tf.inverse()).getOrigin();
>     tf::Quaternion global_q =
>         (sensor_tf * sensor_world_tf * sensor_tf.inverse()).getRotation();
> ```

Save the transformation in a variable to avoid repeating the same computation multiple times

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 57 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7O0-m2QwB1UAeCOiJ:-Kt7O0-m2QwB1UAeCOiK:b-t61gfe) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L57)):*
> ```objc
>         (sensor_tf * sensor_world_tf * sensor_tf.inverse()).getRotation();
> 
>     if (abs(global_pos[0] - data.localpos.x) < config_.max_divergence() ||
> ```

This should be "and" not "or" !!

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 61 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7NUmBjsWT9V1JVBqy:-Kt7NUmBjsWT9V1JVBqz:bvjf1g0) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L61)):*
> ```objc
>         abs(global_pos[2] - data.localpos.z) < config_.max_divergence()) {
>       sensor_status_ = SensorStatus::VALID;
>       bad_data_counter = 0;
> ```

Instead of a counter store the time (ros time). And when data is bad just check the current time -  last good data time is less than timeout. And timeout can be easily configured to be in seconds like 0.2 seconds or something.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 65 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OuRoZfsUKpyFV9AQ:-Kt7OuRoZfsUKpyFV9AR:b-o6zwiy) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L65)):*
> ```objc
>       if (first_msg) {
>         first_msg = false;
>         last_msg_time = msg->header.stamp;
> ```

initialize last_pos

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 68 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OAWtaMzWB85F3GMT:-Kt7OAWtaMzWB85F3GMU:b-32rzu8) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L68)):*
> ```objc
>       } else {
>         ros::Time current_msg_time = msg->header.stamp;
>         double dt = (current_msg_time - last_msg_time).toSec();
> ```

Check if dt is too small and clip dt or skip measurement.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 84 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7QbaV2OuEg-uWx5Ln:-Kt7QbaV2OuEg-uWx5Lo:b-8ikbtl) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L84)):*
> ```objc
>         sensor_data_ = vel_sensor_data;
>       }
>     } else {
> ```

What happens if data is invalid and becomes valid again? You should reset the last message if sensor-status is invalid. You do not need a first message flag. After checking the current sensor validity, check if the sensor was already valid or not. If the sensor is previously invalid and is now valid, set the las_pos, and last_msg_time.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 114 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OxSzRxlrp8Rw8QZt:-Kt7OxSzRxlrp8Rw8QZu:b-t6xgfg) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L114)):*
> ```objc
>   * @brief variable to store last pose
>   */
>   PositionYaw last_pos;
> ```

Make last_pose a tf Vector3

---

*[proto/velocity_sensor_config.proto, line 5 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Pyi2VFg5sn9OBMnd:-Kt7Pyi2VFg5sn9OBMne:b99sciy) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/proto/velocity_sensor_config.proto#L5)):*
> ```protobuf
> message VelocitySensorConfig {
>   /**
>   * @brief transform between sensor and robot
> ```

be more clear. Transform "from" robot center "to" sensor

---

*[proto/velocity_sensor_config.proto, line 9 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7PrWG6vw8VcV4wtsj:-Kt7PrWH2DD_2bP1ZqJo:b-cl3mlk) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/proto/velocity_sensor_config.proto#L9)):*
> ```protobuf
>   repeated double sensor_transform = 2;
>   /**
>   * @brief maximum allowed divergence from gps data
> ```

in meters along any axis

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 9 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Rt-i0Q6zF0dwnGB0:-Kt7Rt-jJZ3oOkPL8JVr:bd8hzpr) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L9)):*
> ```cpp
>   drone_hardware_.getquaddata(quad_data);
> 
>   Joystick joy_data;
> ```

You can construct in the same line as JoyStick joy_data(args);

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 15 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7S3SpaQW4kgN8TcoQ:-Kt7S3SpaQW4kgN8TcoR:b-4c94ee) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L15)):*
> ```cpp
>   VelocityYaw vel_sensor_data = velocity_sensor_.getSensorData();
>   sensor_data = std::make_tuple(joy_data, vel_sensor_data);
>   return true;
> ```

Don't you have to check if the velocity sensor status is valid??

---

*[tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp, line 67 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7SYsFc_QFITLoKbs3:-Kt7SYsFc_QFITLoKbs4:b-vwp4l6) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp#L67)):*
> ```cpp
> }
> 
> int main(int argc, char **argv) {
> ```

Have to add tests for velocity sensor to make sure it is doing the right thing. Also, have to check connector if it does the right thing if velocity sensor goes critical!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 2 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[include/aerial_autonomy/controller_hardware_connectors/joystick_velocity_controller_drone_connector.h, line 35 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7RPV15bdAjvIiaavQ:-Kt7sp6jYORkeljKc5vC:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/controller_hardware_connectors/joystick_velocity_controller_drone_connector.h#L35)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Move implementations to cpp
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 43 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Ks8odr9oSo9I_S8y:-Kt7sprent0obSv1QawW:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/base_sensor.h#L43)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

why should external classes set sensor status? This function should not be public.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/guidance.h, line 18 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7LMaxFZgM7AQ-R93D:-Kt7sq_UI2AFgcf-c75P:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/guidance.h#L18)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

set sensor status to valid always for guidance
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 23 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7QIGiUIPK_6ouwAqo:-Kt7srGU5zDnr6ZGAFlu:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L23)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Ideally, should be taking a GPSSensor as an option and check validity if the sensor is available. Add todo for now
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7MEZjb8EC94IwL9xp:-Kt7ss8DK9NqAOFC2gLl:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L53)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

This is still wrong. The quadrotor takes the world frame aligned with NED (magnetic) of earth and with origin at where quadrotor started. So you need to premultiply with quads initial transformation when it started. In the constructor store the quads zero transformation(Assuming we start the node when quad is landed) and pre multiply your transformation
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 55 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7N8SvK6Zb02iCweT1:-Kt7ssgXDq0uCwhIN4lY:b-3iectd) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L55)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Save the transformation in a variable to avoid repeating the same computation multiple times
</blockquote></details>

sensor_world_tf is tf from sensor data. Its not a constant variable.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 57 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7O0-m2QwB1UAeCOiJ:-Kt7t4aKsV1xFJWSBXmx:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L57)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

This should be "and" not "or" !!
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 61 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7NUmBjsWT9V1JVBqy:-Kt7t5EMiBX8iXuevOmc:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L61)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Instead of a counter store the time (ros time). And when data is bad just check the current time -  last good data time is less than timeout. And timeout can be easily configured to be in seconds like 0.2 seconds or something.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 65 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OuRoZfsUKpyFV9AQ:-Kt7t5fXCzITDna7nXe5:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L65)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

initialize last_pos
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 68 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OAWtaMzWB85F3GMT:-Kt7t654TRq8XPjj54zs:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L68)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Check if dt is too small and clip dt or skip measurement.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 84 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7QbaV2OuEg-uWx5Ln:-Kt7t7U3WExIP0cDdT7i:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L84)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

What happens if data is invalid and becomes valid again? You should reset the last message if sensor-status is invalid. You do not need a first message flag. After checking the current sensor validity, check if the sensor was already valid or not. If the sensor is previously invalid and is now valid, set the las_pos, and last_msg_time.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 114 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7OxSzRxlrp8Rw8QZt:-Kt7t9fCBqs0qDKHpFI2:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L114)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Make last_pose a tf Vector3
</blockquote></details>

Done.

---

*[proto/velocity_sensor_config.proto, line 5 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Pyi2VFg5sn9OBMnd:-Kt7tAEyssVc7X2HXMlr:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/proto/velocity_sensor_config.proto#L5)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

be more clear. Transform "from" robot center "to" sensor
</blockquote></details>

Done.

---

*[proto/velocity_sensor_config.proto, line 9 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7PrWG6vw8VcV4wtsj:-Kt7tAjHEpmgEnoAD8Ku:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/proto/velocity_sensor_config.proto#L9)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

in meters along any axis
</blockquote></details>

Done.

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 9 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7Rt-i0Q6zF0dwnGB0:-Kt7tBWdkFHw9mFxru1U:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L9)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

You can construct in the same line as JoyStick joy_data(args);
</blockquote></details>

Done.

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 15 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7S3SpaQW4kgN8TcoQ:-Kt7tBzuYgCFs2-kKwBp:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L15)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Don't you have to check if the velocity sensor status is valid??
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 25 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7vSUbLZ-WsyXJ6tim:-Kt7vSUbLZ-WsyXJ6tin:blml3yb) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/base_sensor.h#L25)):*
> ```objc
>   * @brief Constructor
>   */
>   Sensor() : sensor_status_(SensorStatus::INVALID) {}
> ```

Why did you change this implementation??

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 25 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7w-L6tyaB_oYTspvz:-Kt7w-L6tyaB_oYTspw-:bd31695) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/base_sensor.h#L25)):*
> ```objc
>   * @brief Constructor
>   */
>   Sensor() {
> ```

Create a constructor with initial sensor status and make that constructor protected.

as in
```
protected:

 Sensor(SensorStatus sensor_status): sensor_status_(sensor_status){}
```

---

*[include/aerial_autonomy/sensors/guidance.h, line 19 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7vZMkzY3oOPdmAXP1:-Kt7vZMkzY3oOPdmAXP2:b-26gyjr) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/guidance.h#L19)):*
> ```objc
>   */
>   Guidance(parsernode::Parser &drone_hardware)
>       : drone_hardware_(drone_hardware) {
> ```

just use parent constructor Sensor(SensorStatus::VALID) based on above comment.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 55 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7N8SvK6Zb02iCweT1:-Kt7x1xNjATVA2BVYW2Z:bp72c2d) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L55)):*
<details><summary><i>Previously, sohampatkar10 wrote…</i></summary><blockquote>

sensor_world_tf is tf from sensor data. Its not a constant variable.
</blockquote></details>

save quad_initial_tf* sensor_tf * sensor_world_tf*sensor_tf.inverse() in a transform variable and global_pos, global_q can be obtained from this transform.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 57 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7O0-m2QwB1UAeCOiJ:-Kt7xFm26jDV024ck1WN:by2ac2l) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L57)):*
<details><summary><i>Previously, sohampatkar10 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Write tests to explicitly capture this logic. Show that global_pos[0]  being close to data.localpos.x is not enough to make the sensor data valid.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 91 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7yLP7tmQUHIJcmqwM:-Kt7yLP7tmQUHIJcmqwN:by20ymo) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/velocity_sensor.h#L91)):*
> ```objc
>         vel_sensor_data.z = (global_pos[2] - last_pos.z) / dt;
> 
>         last_pos = global_pos;
> ```

This and the next line are repeated in both if and else conditions. You can move these two lines out of the if/else loop.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 93 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7ydPmZpSjHMKswCI8:-Kt7ydPmZpSjHMKswCI9:byxqh9p) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/velocity_sensor.h#L93)):*
> ```objc
>         last_pos = global_pos;
>         last_msg_time = current_msg_time;
>         last_good_data_time = current_msg_time;
> ```

As far as I understand the problem, you do not need last_good_data_time. last_msg_time is only specified if the data is valid. You do not compute the velocity even if the current sensor data is valid if your previous sensor data is invalid. So just use last_msg_time for differentiating velocity.

Also for checking if sensor data is valid or not just use last_msg_time. Essentially, last_msg_time is the same as last_good_data_time.

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 30 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7zpPehI5WP_vwe6Z0:-Kt7zpPfuYnHM_aIMEDC:b-idgzxd) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L30)):*
> ```cpp
>     sensor_data = std::make_tuple(joy_data, vel_sensor_data);
>     return true;
>   }
> ```

If possible just return bool(velocity_sensor_.getSensorStatus()) instead of multiple return statements.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 4 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 25 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7vSUbLZ-WsyXJ6tim:-Kt8pk6Q3_j4Pukvqa1u:bbbk1cn) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/base_sensor.h#L25)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Why did you change this implementation??
</blockquote></details>

It was throwing some error. I made the new constructor that seems to be working.  

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 25 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7w-L6tyaB_oYTspvz:-Kt8pwqow_FPaa_2moSf:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/base_sensor.h#L25)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Create a constructor with initial sensor status and make that constructor protected.

as in
```
protected:

 Sensor(SensorStatus sensor_status): sensor_status_(sensor_status){}
```
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/guidance.h, line 19 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7vZMkzY3oOPdmAXP1:-Kt8q-g2-lQ-AM_58AhS:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/guidance.h#L19)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

just use parent constructor Sensor(SensorStatus::VALID) based on above comment.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 55 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7N8SvK6Zb02iCweT1:-Kt8q3TJvoivfeJTbUCb:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L55)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

save quad_initial_tf* sensor_tf * sensor_world_tf*sensor_tf.inverse() in a transform variable and global_pos, global_q can be obtained from this transform.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 57 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7O0-m2QwB1UAeCOiJ:-Kt8q4f4zPALCei8oiSe:bpjapb6) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L57)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Write tests to explicitly capture this logic. Show that global_pos[0]  being close to data.localpos.x is not enough to make the sensor data valid.
</blockquote></details>

I will add a VelocitySensor rostest and check this. 

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 91 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7yLP7tmQUHIJcmqwM:-Kt8qCqv_7g4YkWnKdxE:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/velocity_sensor.h#L91)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

This and the next line are repeated in both if and else conditions. You can move these two lines out of the if/else loop.
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 93 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7ydPmZpSjHMKswCI8:-Kt8qDa6d59GNfdNvhhd:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/include/aerial_autonomy/sensors/velocity_sensor.h#L93)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

As far as I understand the problem, you do not need last_good_data_time. last_msg_time is only specified if the data is valid. You do not compute the velocity even if the current sensor data is valid if your previous sensor data is invalid. So just use last_msg_time for differentiating velocity.

Also for checking if sensor data is valid or not just use last_msg_time. Essentially, last_msg_time is the same as last_good_data_time.
</blockquote></details>

Done.

---

*[src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp, line 30 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7zpPehI5WP_vwe6Z0:-Kt8qGH_XWcGAd1xrWjm:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/957288e303621f8b5f45281975f3f76b7aee6ca0/src/controller_hardware_connectors/joystick_velocity_controller_drone_connector.cpp#L30)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

If possible just return bool(velocity_sensor_.getSensorStatus()) instead of multiple return statements.
</blockquote></details>

Done.

---

*[tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp, line 67 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7SYsFc_QFITLoKbs3:-Kt8qHUx_LmLMgStCQmi:b4oa5tt) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp#L67)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Have to add tests for velocity sensor to make sure it is doing the right thing. Also, have to check connector if it does the right thing if velocity sensor goes critical!
</blockquote></details>

I will add a VelocitySensor rostest and check all this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 1 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 57 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7O0-m2QwB1UAeCOiJ:-KtAqD26tsF3qzsudW4A:b7514is) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/include/aerial_autonomy/sensors/velocity_sensor.h#L57)):*
<details><summary><i>Previously, sohampatkar10 wrote…</i></summary><blockquote>

I will add a VelocitySensor rostest and check this. 
</blockquote></details>

Added VelocitySensor rostest

---

*[tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp, line 67 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-Kt7SYsFc_QFITLoKbs3:-KtAq2yVW4ZC1blh7tQ7:b91vf34) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/e0b8920df286ed818c3619198b3601f5b74dcd47/tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp#L67)):*
<details><summary><i>Previously, sohampatkar10 wrote…</i></summary><blockquote>

I will add a VelocitySensor rostest and check all this. 
</blockquote></details>

Added VelocitySensor ros test. Also added test to check behaviour of controller connector when sensor status is critical.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Reviewed 2 of 4 files at r3, 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 13 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCaPAi3JhX6aiHouHC:-KtCaPAi3JhX6aiHouHD:bstkvc3) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/include/aerial_autonomy/sensors/base_sensor.h#L13)):*
> ```objc
> * for a prolonged period
> */
> enum class SensorStatus { INVALID = 0, VALID = 1 };
> ```

Can you add a function here that converts SensorStatus to bool explicitly

---

*[proto/velocity_sensor_config.proto, line 19 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtC_aBuYS4VKQn9_mL0:-KtC_aBuYS4VKQn9_mL1:b-eq7lp0) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/proto/velocity_sensor_config.proto#L19)):*
> ```protobuf
>   optional double bad_data_timeout = 4 [default = 0.5];
>   /**
>   * @brief minimum timestep in seconds 
> ```

Can you add more info. I have no clue what minimum timestep means.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 22 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbzKZ57ZjcdpTSYQt:-KtCbzKZ57ZjcdpTSYQu:b-7i0euv) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L22)):*
> ```cpp
>     pose_pub.publish(pose);
>     ros::Duration(0.01).sleep();
>     ros::spinOnce();
> ```

Do not need sleep and ros spin here.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 46 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbuoQOG4o_-pasLuN:-KtCbuoQOG4o_-pasLuO:b-fh5co4) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L46)):*
> ```cpp
>   pose.pose.orientation.w = 1.0;
>   publishData(pose);
>   ASSERT_TRUE(bool(sensor.getSensorStatus()));
> ```

use waitUntilTrue with ros spinOnce inside

---

*[tests/sensors/velocity_sensor_tests.cpp, line 53 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCc2bjNrXGonMK_bDA:-KtCc2bjNrXGonMK_bDB:bwz10zz) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L53)):*
> ```cpp
>   geometry_msgs::PoseStamped pose;
>   pose.header.stamp = ros::Time::now();
>   pose.pose.position.x = 0.6;
> ```

I do not understand why the sensor data is invalid here and is valid above?

---

*[tests/sensors/velocity_sensor_tests.cpp, line 98 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbTGiuKPtvLdDG0j-:-KtCbTGiuKPtvLdDG0j0:b9ntdgc) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L98)):*
> ```cpp
>   VelocitySensor sensor(drone_hardware, nh, config);
>   geometry_msgs::PoseStamped pose;
>   for (int i = 0; i > 40; ++i) {
> ```

i < 40??

---

*[tests/sensors/velocity_sensor_tests.cpp, line 120 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcNohgrTnxoKXuW6g:-KtCcNohgrTnxoKXuW6h:b-l0q7tv) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L120)):*
> ```cpp
>   new_config.add_sensor_transform(0.0);
> 
>   tf::Transform sensor_tf;
> ```

can construct in same line...

---

*[tests/sensors/velocity_sensor_tests.cpp, line 126 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcgV8-_PNnmT9tw2F:-KtCcgV8-_PNnmT9tw2G:b-bo6zw6) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L126)):*
> ```cpp
>   VelocitySensor sensor(drone_hardware, nh, new_config);
>   ros::Time t0 = ros::Time::now();
>   ros::Duration(0.03).sleep();
> ```

You can do `ros::Time t1 = t0 + ros::Duration(0.03)`. You do not need to sleep to create a new time!

---

*[tests/sensors/velocity_sensor_tests.cpp, line 148 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcXrDNozU-7AqnEUA:-KtCcXrDNozU-7AqnEUB:besarqu) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L148)):*
> ```cpp
>   publishData(pose);
> 
>   double dt = (t1 - t0).toSec();
> ```

Why do you need t0?? Is this not 0.03 directly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[include/aerial_autonomy/sensors/base_sensor.h, line 13 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCaPAi3JhX6aiHouHC:-KtDIsRPceaWzsTabRUh:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/include/aerial_autonomy/sensors/base_sensor.h#L13)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Can you add a function here that converts SensorStatus to bool explicitly
</blockquote></details>

Done.

---

*[proto/velocity_sensor_config.proto, line 19 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtC_aBuYS4VKQn9_mL0:-KtDIxYqjah4ohRYm5nJ:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/proto/velocity_sensor_config.proto#L19)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Can you add more info. I have no clue what minimum timestep means.
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 22 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbzKZ57ZjcdpTSYQt:-KtDJ03LlMZ_kKl4J4rw:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L22)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Do not need sleep and ros spin here.
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 46 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbuoQOG4o_-pasLuN:-KtDJ1h5aFPzxkPmajNO:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L46)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

use waitUntilTrue with ros spinOnce inside
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 53 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCc2bjNrXGonMK_bDA:-KtDJ630DJLv9vQmCKhu:bkniyys) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L53)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

I do not understand why the sensor data is invalid here and is valid above?
</blockquote></details>

One drifts from quad data more than the other. Added way point cmd to make it explicit. 

---

*[tests/sensors/velocity_sensor_tests.cpp, line 98 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCbTGiuKPtvLdDG0j-:-KtDJOFYwJz4WohDu5C_:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L98)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

i < 40??
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 120 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcNohgrTnxoKXuW6g:-KtDJQ_jMf09sF9YL_oo:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L120)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

can construct in same line...
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 126 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcgV8-_PNnmT9tw2F:-KtDJSvzbA9RhMppU47W:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L126)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

You can do `ros::Time t1 = t0 + ros::Duration(0.03)`. You do not need to sleep to create a new time!
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 148 at r4](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtCcXrDNozU-7AqnEUA:-KtDJUlj9rstLM68B3jQ:bi63ngk) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/1a246ee483b89e947a49db8d974d12f8d0c40b9d/tests/sensors/velocity_sensor_tests.cpp#L148)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Why do you need t0?? Is this not 0.03 directly?
</blockquote></details>

It wasnt when I was using ros::sleep. Changed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 126 at r5](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtErHVbFrB5AKfNnQZe:-KtErHVbFrB5AKfNnQZf:bgh0p3d) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/bbd898a280f770187a65d72e195dd39adad327a1/tests/sensors/velocity_sensor_tests.cpp#L126)):*
> ```cpp
>   quad_pos.y = -0.2;
>   quad_pos.z = 0.3;
>   drone_hardware.cmdwaypoint(quad_pos);
> ```

Start with sensor status as true by publishing a valid pose. 
Check that sensor status is VALID
Then in your sensor_status_check function, keep publishing invalid pose until sensor status becomes invalid.

No need for a separate for loop with some arbitrary number like 40. 

---

*[tests/sensors/velocity_sensor_tests.cpp, line 147 at r5](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtEs3EZXt_V0LEE4H-x:-KtEs3E_O39c5Lz0KO6v:b-207z1m) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/bbd898a280f770187a65d72e195dd39adad327a1/tests/sensors/velocity_sensor_tests.cpp#L147)):*
> ```cpp
> }
> 
> TEST_F(VelocitySensorTests, SensorTF) {
> ```

Add a test for minimum timestep. If poses are published too quickly, is the velocity computed after clipping time step.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 126 at r5](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtErHVbFrB5AKfNnQZe:-KtF05425z9eu-DINol7:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/bbd898a280f770187a65d72e195dd39adad327a1/tests/sensors/velocity_sensor_tests.cpp#L126)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Start with sensor status as true by publishing a valid pose. 
Check that sensor status is VALID
Then in your sensor_status_check function, keep publishing invalid pose until sensor status becomes invalid.

No need for a separate for loop with some arbitrary number like 40. 
</blockquote></details>

Done.

---

*[tests/sensors/velocity_sensor_tests.cpp, line 147 at r5](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtEs3EZXt_V0LEE4H-x:-KtF030Cd8yi2ex64Uap:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/bbd898a280f770187a65d72e195dd39adad327a1/tests/sensors/velocity_sensor_tests.cpp#L147)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Add a test for minimum timestep. If poses are published too quickly, is the velocity computed after clipping time step.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Reviewed 1 of 6 files at r2, 2 of 4 files at r3, 4 of 7 files at r4, 2 of 3 files at r5.
Review status: 9 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 41 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF8VtsrogsDQihFwg2:-KtF8VtsrogsDQihFwg3:b-tch21v) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L41)):*
> ```objc
>         data.rpydata.x, data.rpydata.y, data.rpydata.z);
>     quad_intial_tf.setOrigin(quad_intial_pos);
>     quad_intial_tf.setRotation(quad_intial_rot);
> ```

can just create the Vector3 and Quaternion directly as the argument instead of using a temp variable

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 61 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF4-6MsyQOH_JqdkBq:-KtF4-6NIfh_jd-0bDdO:b-oyf0q6) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L61)):*
> ```objc
>                        msg->pose.orientation.z, msg->pose.orientation.w);
> 
>     tf::Transform sensor_world_tf;
> ```

can do this in fewer lines, e.g.
`tf::Transform sensor_world_tf(tf::Quaternion(msg->pose.orientation.x, msg->pose.orientation.y,msg->pose.orientation.z, msg->pose.orientation.w), tf::Vector3(msg->pose.position.x, msg->pose.position.y, msg->pose.position.z));`

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 66 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF8vEPVdrgg-o2xTYi:-KtF8vEPVdrgg-o2xTYj:bvsoxo8) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L66)):*
> ```objc
> 
>     tf::Transform global_tf =
>         quad_intial_tf * sensor_tf * sensor_world_tf * sensor_tf.inverse();
> ```

Can you comment on this.  What is global_tf and how is it computed? 

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 119 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF3f0W9-ZlE2awnAOZ:-KtF3f0Xp1dn1WyXVpSl:b-7qj5w8) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L119)):*
> ```objc
>   * @brief Transform between sensor and robot
>   */
>   tf::Transform sensor_tf;
> ```

class variable names should end with _

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 123 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF3pwYz4a2h6r_YisG:-KtF3pwYz4a2h6r_YisH:b-39ryox) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L123)):*
> ```objc
>   * @brief quad's initial transform in world (NED) frame
>   */
>   tf::Transform quad_intial_tf;
> ```

initial

---

*[tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp, line 67 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtFFAoKsi06WZBYdiFS:-KtFFAoKsi06WZBYdiFT:b7jqiz3) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp#L67)):*
> ```cpp
> }
> 
> TEST(JoystickVelocityControllerDroneConnector, SensorCritical) {
> ```

What is making the sensor critical here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

sohampatkar10 commented:
Review status: 8 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 41 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF8VtsrogsDQihFwg2:-KtFQWxw2y2dVSuz1hR7:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L41)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

can just create the Vector3 and Quaternion directly as the argument instead of using a temp variable
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 61 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF4-6MsyQOH_JqdkBq:-KtFQYGt6SY-wsIX2XFV:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L61)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

can do this in fewer lines, e.g.
`tf::Transform sensor_world_tf(tf::Quaternion(msg->pose.orientation.x, msg->pose.orientation.y,msg->pose.orientation.z, msg->pose.orientation.w), tf::Vector3(msg->pose.position.x, msg->pose.position.y, msg->pose.position.z));`
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 66 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF8vEPVdrgg-o2xTYi:-KtFQZnLqaPUntBvyv9I:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L66)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Can you comment on this.  What is global_tf and how is it computed? 
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 119 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF3f0W9-ZlE2awnAOZ:-KtFQb0totU4iOaR4FmA:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L119)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

class variable names should end with _
</blockquote></details>

Done.

---

*[include/aerial_autonomy/sensors/velocity_sensor.h, line 123 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtF3pwYz4a2h6r_YisG:-KtFQcafr9CwgH0AAR_W:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/include/aerial_autonomy/sensors/velocity_sensor.h#L123)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

initial
</blockquote></details>

Done.

---

*[tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp, line 67 at r6](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102#-KtFFAoKsi06WZBYdiFS:-KtFQfG63tO4K-HORDqT:b-2pztio) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/0ca2dc350befe956c62151fe30cd4f03edec5a65/tests/controller_hardware_connectors/joystick_velocity_controller_drone_connector_tests.cpp#L67)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

What is making the sensor critical here?
</blockquote></details>

Sensor Status is invalid by default unless we use the Sensor(status) type of constructor. Guidance sensor is always Valid so it works for rest of the tests. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/102)*
<!-- Sent from Reviewable.io -->
